sdcm/sct_events.py

```diff
@@ -0,0 +1,270 @@
+import logging
+import zmq
+import json
+from json import JSONEncoder
+import errno
+import os
+import signal
+import datetime
+
+from multiprocessing import Process, Value, Event
+
+import enum
+from enum import Enum
+
+LOGGER = logging.getLogger(__name__)
+
+
+class EventsDevice(Process):
+
+    def __init__(self):
+        super(EventsDevice, self).__init__()
+        self.ready_event = Event()
+        self.pub_port = Value('d', 0)
+        self.sub_port = Value('d', 0)
+
+    def run(self):
+        try:
+            context = zmq.Context(1)
+            # Socket facing clients
+            frontend = context.socket(zmq.SUB)
+            self.pub_port.value = frontend.bind_to_random_port("tcp://*")
+            frontend.setsockopt(zmq.SUBSCRIBE, "")
+
+            # Socket facing services
+            backend = context.socket(zmq.PUB)
+            self.sub_port.value = backend.bind_to_random_port("tcp://*")
+            LOGGER.info("EventDevice Listen on pub_port=%d, sub_port=%d", self.pub_port.value, self.sub_port.value)
+
+            backend.setsockopt(zmq.LINGER, 0)
+            frontend.setsockopt(zmq.LINGER, 0)
+
+            self.ready_event.set()
+            zmq.proxy(frontend, backend)
+
+        except Exception as e:
+            LOGGER.exception("zmq device failed")
+        finally:
+            frontend.close()
+            backend.close()
+            context.term()
+
+    def subscribe_events(self, filter_type=''):
+        context = zmq.Context()
+        LOGGER.info("subscribe to server with port %d", self.sub_port.value)
+        socket = context.socket(zmq.SUB)
+        socket.connect("tcp://localhost:%d" % self.sub_port.value)
+        socket.setsockopt(zmq.SUBSCRIBE, filter_type)
+
+        while True:
+            obj = socket.recv_pyobj()
+            LOGGER.debug("got event_type=%s, obj=%s", obj.__class__.__name__, obj)
+            yield obj.__class__.__name__, obj
+
+    def publish_event(self, event):
+        context = zmq.Context()
+        socket = context.socket(zmq.PUB)
+        socket.connect("tcp://localhost:%d" % self.pub_port.value)
+        time.sleep(0.01)
+
+        socket.send_pyobj(event)
+
+
+EVENTS_PROCESS = EventsDevice()
+
+# monkey patch JSONEncoder make enums jsonable
+_saved_default = JSONEncoder().default  # Save default method.
+
+
+def _new_default(self, obj):
+    if isinstance(obj, Enum):
+        return obj.name  # Could also be obj.value
+    else:
+        return _saved_default
+
+
+JSONEncoder.default = _new_default  # Set new default method.
+
+
+class Priority(enum.Enum):
```

Comment:
Priority -> Severity

https://www.softwaretestingmaterial.com/what-is-the-difference-between-severity-and-priority-in-software-testing/